kvclient: clean up NLHE handling

andrewbaptist opened the pull request:
Previously we could reprocess a NLHE even if we didn't update our
internal sequence. This change better detects if there is an actual
leaseholder change and additionally updates the `leaseholder` variable
which is used in other places in the function and should be correct.

Epic: none

Release note: None

Commit:
kvclient: allow maybeUpdate to take empty leases

Previously SyncTokenAndMaybeUpdateCache would throw an error if it was
called with a nil Lease or empty RangeDescriptor. This prevented use in
certain cases where the server didn't have the lease information.
This change allows calling the method with an empty lease.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/119747)
<!-- Reviewable:end -->

Commit:
kvclient: clean up NLHE handling

Previously we could reprocess a NLHE even if we didn't update our
internal sequence. This change better detects if there is an actual
leaseholder change and additionally updates the `leaseholder` variable
which is used in other places in the function and should be correct.

Epic: none

Release note: None

andrewbaptist commented:
@arulajmani 

This change helps clean up the NLHE handling. This becomes more important for proxied requests where there are now 4 sources of information:
1) Original client
2) DistSender of the proxy node
3) Replica information of the proxy node
4) Replica information of the leaseholder

In general they could all be out of sync, and it becomes important to tighten up the checks here in preparation for that dropping. This change additionally more correctly handles a case where there is both a range change and a different leaseholder (for instance after a split / merge) by detecting the invalid token and immediately exiting with a SendError rather than continuing on with additional replicas.

arulajmani reviewed (changes requested):
> This change helps clean up the NLHE handling. This becomes more important for proxied requests where there are now 4 sources of information:

Why would proxied requests return NLHE errors back to the original client? Wouldn't they either:
1. Figure out who the leaseholder is, route to it, and return the result.
2. OR in some rare cases, try all replicas (maybe after one round of range descriptor eviction), not be able to serve the request, and return an error back to the original client?

----

Separately, after reviewing this change, it looks like we're generalizing some of the assumptions that were being made by the previous code about what it considered valid input. I think it's valuable to keep the assumptions in place instead of making things work with all possible inputs -- it means we have to be intentional with our changes in the future and not have to account for impossible cases while debugging this area in the future. 

Can we simplify this PR to just get rid of the `updatedLeaseholder` return value on the production path for `SyncTokenAndMaybeUpdateCache`?

> Reviewed 1 of 1 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-:-NrqE4U8-GzddEeC1xu_:b-kkxtey)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2670 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqJXdZ99CPKQDINyfD:-NrqJXdZ99CPKQDINyfE:b-6c7xuy) ([raw file](https://github.com/cockroachdb/cockroach/blob/903e2fc5e8286ecff775bd7407bd399889ca97ce/pkg/kv/kvclient/kvcoord/dist_sender.go#L2670)):*
> ```Go
> 				// If we got some lease information, we use it. If not, we loop around
> 				// and try the next replica.
> 				if tErr.Lease != nil {
> ```

Do we need to get rid of this structure? There's nothing we can do if the NLHE doesn't point to a new lease, so trying the next replica makes sense. 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2676 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqJyx1BuI1-0DErr25:-NrqJyx1BuI1-0DErr26:b-vp5wsz) ([raw file](https://github.com/cockroachdb/cockroach/blob/903e2fc5e8286ecff775bd7407bd399889ca97ce/pkg/kv/kvclient/kvcoord/dist_sender.go#L2676)):*
> ```Go
> 
> 					var updatedLeaseholder bool
> 					if tErr.Lease != nil {
> ```

This if condition from before seems unnecessary, let's remove it.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2683 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqK8zl6eBtgD9IIMaS:-NrqK8zl6eBtgD9IIMaT:b-ipsruw) ([raw file](https://github.com/cockroachdb/cockroach/blob/22af9b4fbf7095f22da55da414285e6c8a05a645/pkg/kv/kvclient/kvcoord/dist_sender.go#L2683)):*
> ```Go
> 				}
> 				routing.SyncTokenAndMaybeUpdateCache(ctx, tErr.Lease, &tErr.RangeDesc)
> 				// If the range bounds changed our routing may be invalid.
> ```

Wouldn't this have been caught by `skipStaleReplicas`? Is there benefit to adding a special case here, as opposed to checking for invalid routing in just a single place?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2692 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqKO_lCltti2vbtbxv:-NrqKO_lCltti2vbtbxw:b-g6grzw) ([raw file](https://github.com/cockroachdb/cockroach/blob/22af9b4fbf7095f22da55da414285e6c8a05a645/pkg/kv/kvclient/kvcoord/dist_sender.go#L2692)):*
> ```Go
> 				// If we got new information, we use it. If not, we loop around and try
> 				// the next replica.
> 				if leaseholder != nil && tErr.Lease != nil {
> ```

We won't need this change if we don't remove the `tErr.Lease != nil` branch above, right?
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 400 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqMYcR1spKrEA_hwQg:-NrqMYcS-K7fJXXR6U9T:bb9olye) ([raw file](https://github.com/cockroachdb/cockroach/blob/505dcae125dd4b24758f87968b405b90dd4d7b75/pkg/kv/kvclient/rangecache/range_cache.go#L400)):*
> ```Go
> // cached.
> //
> // TODO(arul): Instead of returning updatedLeaseholder all the way back to the
> ```

Thanks for addressing this! 
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1460 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqGkkK8Ya9TN6M33zu:-NrqGkkK8Ya9TN6M33zv:bf3dkz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/505dcae125dd4b24758f87968b405b90dd4d7b75/pkg/kv/kvclient/rangecache/range_cache.go#L1460)):*
> ```Go
> 	ctx context.Context, l *roachpb.Lease, rangeDesc *roachpb.RangeDescriptor,
> ) (updated, updatedLease bool, newEntry *cacheEntry) {
> 	if rangeDesc.RangeID != 0 && !descsCompatible(&e.desc, rangeDesc) {
> ```

When would this be called with an uninitialized descriptor? More importantly, it's a bit subtle to allow comparing uninitialized descriptors with initialized ones using generation numbers below. Unless there's a good reason to, fataling here seems okay.
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1487 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqHgaD5-zosqmAHJKF:-NrqHgaD5-zosqmAHJKG:bc6hev2) ([raw file](https://github.com/cockroachdb/cockroach/blob/505dcae125dd4b24758f87968b405b90dd4d7b75/pkg/kv/kvclient/rangecache/range_cache.go#L1487)):*
> ```Go
> 	// Lastly, if the cached lease is empty, it will be updated with the supplied
> 	// lease regardless of the associated range descriptor's generation.
> 	if l != nil && ((l.Sequence != 0 && e.lease.Sequence != 0 && l.Sequence > e.lease.Sequence) ||
> ```

Similar to the comment above, do we expect this function to be called with a nil lease? Before r2 we never called this with a nil lease -- which makes sense, if the NLHE is not pointing to a lease, there's no "SyncTokenAndMaybeUpdateCache" to do. 

We should be explicit about this contract, and maybe add a log.Fatal instead of an implied runtime panic instead.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> > This change additionally more correctly handles a case where there is both a range change and a different leaseholder (for instance after a split / merge) by detecting the invalid token and immediately exiting with a SendError rather than continuing on with additional replicas.

That should result in a `RangeKeyMismatchError`, right?

EDIT: I meant to delete this, I'd incorrectly assumed this. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-:-NrqNdXl2qwmRgoka9Un:bauu3vq)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist)



<!-- Sent from Reviewable.io -->

andrewbaptist reviewed (commented):
I think the main reason for this change is "what is the meaning of an NLHE". There are a few possible cases it is returned today:

Specifically there is a grid of 2 x 3 x 4 = 24 cases where a NLHE is returned. I don't think they are all well tested or optimally handled (in some cases we throw away information we learned from the server)

IsServerCurrentLeaseholder  - yes, no
IsRangeInfoSame - yes, client newer, server newer
IsLeaseInfoSame - both have same lease info, neither has lease info, client has newer info, server has newer info

In some of these cases we can return a `RangeMismatchError`, but in others we can get back a nil Lease from the client, even if there is different range info in it. My goal of this PR was to not have to individually branch on all these cases, but instead allow the Sync call to be made in all cases.

With the proxy changes that are upcoming, there are additional places where we call `SyncTokenAndMaybeUpdateCache` since the Proxy request will also carry the clients Lease information. So in this case the Lease information is passed to the server and can update the server prior to it issuing the proxy information. This adds a new path for updating the cache related to lease information only. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-:-Nrqtekh7JyFMuMoat6O:b-i28t72)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2670 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqJXdZ99CPKQDINyfD:-Nrqqs_z2N8DUtzFuh-C:bgkgmf1) ([raw file](https://github.com/cockroachdb/cockroach/blob/903e2fc5e8286ecff775bd7407bd399889ca97ce/pkg/kv/kvclient/kvcoord/dist_sender.go#L2670)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Do we need to get rid of this structure? There's nothing we can do if the NLHE doesn't point to a new lease, so trying the next replica makes sense.
</blockquote></details>

The NLHE can contain both a new Lease and a different RangeDesc (for instance after a split). There is the following case:
Range 1 (A-Z)
Splits to Range 1 (A-M), and Range 2 (N-Z)
A request comes in for Range 1 on a non-leaseholder who doesn't know the new leaseholder.
The store responds with a NLHE error with the RangeDesc, but a nil Lease

In that case we still want to call `SyncTokenAndMaybeUpdateCache` with a nil lease.  
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2676 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqJyx1BuI1-0DErr25:-Nrqx1TBCOus0Tc8NVNy:beqd01c) ([raw file](https://github.com/cockroachdb/cockroach/blob/903e2fc5e8286ecff775bd7407bd399889ca97ce/pkg/kv/kvclient/kvcoord/dist_sender.go#L2676)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

This if condition from before seems unnecessary, let's remove it.
</blockquote></details>

Unfortunately there are places like this: https://github.com/cockroachdb/cockroach/blob/b3208e896778c4850704dee8f252141ce6edc2f1/pkg/kv/kvserver/replica_range_lease.go#L386 where we pass back a nil Lease. In this case we would have to make larger changes to fully remove it. 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2683 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqK8zl6eBtgD9IIMaS:-NrqxNcd11MTIhzuC_mt:b-g0562d) ([raw file](https://github.com/cockroachdb/cockroach/blob/22af9b4fbf7095f22da55da414285e6c8a05a645/pkg/kv/kvclient/kvcoord/dist_sender.go#L2683)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Wouldn't this have been caught by `skipStaleReplicas`? Is there benefit to adding a special case here, as opposed to checking for invalid routing in just a single place?
</blockquote></details>

Yes. Changing that to a continue instead is cleaner. In general it might be a better change to make `SyncTokenAndMaybeUpdateCache` return an err if the token is no longer valid. Thoughts? 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2692 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqKO_lCltti2vbtbxv:-NrqzMEL8HuTxTFBxzEa:bkeqeu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/22af9b4fbf7095f22da55da414285e6c8a05a645/pkg/kv/kvclient/kvcoord/dist_sender.go#L2692)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

We won't need this change if we don't remove the `tErr.Lease != nil` branch above, right?
</blockquote></details>

See `TestNoBackoffOnNotLeaseHolderErrorWithoutLease`. There is a valid case where we don't provide a lease, but still return a NLHE. 
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 400 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqMYcR1spKrEA_hwQg:-NrqycmP8vFu3ygo3vbv:b-qbhvpz) ([raw file](https://github.com/cockroachdb/cockroach/blob/505dcae125dd4b24758f87968b405b90dd4d7b75/pkg/kv/kvclient/rangecache/range_cache.go#L400)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Thanks for addressing this!
</blockquote></details>

NP. I'm adding another commit to change the return value to be an error (if the token is no longer valid). That forces future callers to handle that case. 
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1460 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqGkkK8Ya9TN6M33zu:-Nrqywdl8bN6rPTdB1d_:bdsm0s4) ([raw file](https://github.com/cockroachdb/cockroach/blob/505dcae125dd4b24758f87968b405b90dd4d7b75/pkg/kv/kvclient/rangecache/range_cache.go#L1460)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

When would this be called with an uninitialized descriptor? More importantly, it's a bit subtle to allow comparing uninitialized descriptors with initialized ones using generation numbers below. Unless there's a good reason to, fataling here seems okay.
</blockquote></details>

I think I can remove the nil rangeDesc check, but I can't remove the Lease check below. 
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1487 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqHgaD5-zosqmAHJKF:-NrqsDrLEHocKwChxwZK:bocvr41) ([raw file](https://github.com/cockroachdb/cockroach/blob/505dcae125dd4b24758f87968b405b90dd4d7b75/pkg/kv/kvclient/rangecache/range_cache.go#L1487)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Similar to the comment above, do we expect this function to be called with a nil lease? Before r2 we never called this with a nil lease -- which makes sense, if the NLHE is not pointing to a lease, there's no "SyncTokenAndMaybeUpdateCache" to do. 

We should be explicit about this contract, and maybe add a log.Fatal instead of an implied runtime panic instead.
</blockquote></details>

The idea was to allow calling this method when we don't have a new Lease to provide. I agree that it makes the conditions for this a little more subtle and I don't love it, but the alternative was to have seperate methods for `SyncLease` and `SyncRangeDescriptor` Maybe we should talk briefly on this.


<!-- Sent from Reviewable.io -->

andrewbaptist commented:
@arulajmani TFTR. let me know if you want to chat about this PR. 

Thanks!
Andrew

arulajmani reviewed (commented):
Thanks for explaining the context here, I understand the motivation for the added complexity a bit more now. The patch looks fine if we're convinced the complexity is worth it -- can we add a test to demonstrate the case we're doing better in before merging?

> Reviewed 1 of 2 files at r3, 1 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-:-Nrux6ybEeOqPgXT9HFJ:b-ysudq1)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2670 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqJXdZ99CPKQDINyfD:-NrusetJBhSI9Z_jL6Lv:bskevm5) ([raw file](https://github.com/cockroachdb/cockroach/blob/903e2fc5e8286ecff775bd7407bd399889ca97ce/pkg/kv/kvclient/kvcoord/dist_sender.go#L2670)):*
> A request comes in for Range 1 on a non-leaseholder who doesn't know the new leaseholder.

Getting a nil lease here is extremely rare (I'm looking at the callers of `NewNotLeaseHolderError`). So there aren't many practical cases where we'd have both a nil lease and a stale range descriptor. Do we think the theoretical benefit here is worth the added complexity?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2676 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqJyx1BuI1-0DErr25:-NrurUeaDfNWrjE96vwR:b-curhc6) ([raw file](https://github.com/cockroachdb/cockroach/blob/903e2fc5e8286ecff775bd7407bd399889ca97ce/pkg/kv/kvclient/kvcoord/dist_sender.go#L2676)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Unfortunately there are places like this: https://github.com/cockroachdb/cockroach/blob/b3208e896778c4850704dee8f252141ce6edc2f1/pkg/kv/kvserver/replica_range_lease.go#L386 where we pass back a nil Lease. In this case we would have to make larger changes to fully remove it.
</blockquote></details>

Sorry for not being clear, this was contingent on my previous comment of keeping the old structure -- the if condition here isn't necessary (on master) because it's already inside the same if condition above.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2683 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqK8zl6eBtgD9IIMaS:-NruvAg12mkhnA3LUaSX:b35uhtj) ([raw file](https://github.com/cockroachdb/cockroach/blob/22af9b4fbf7095f22da55da414285e6c8a05a645/pkg/kv/kvclient/kvcoord/dist_sender.go#L2683)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Yes. Changing that to a continue instead is cleaner. In general it might be a better change to make `SyncTokenAndMaybeUpdateCache` return an err if the token is no longer valid. Thoughts?
</blockquote></details>

We're now assuming that the only error here means the routing is invalid, which might not hold true forever. One pattern I've seen used before is to return a special type of error (say `routingInvalidError`), check for that, and only continue in that case. 

However, for this, it might just be easier to check if the routing is valid or not, instead of introducing a new error type. Alternatively, I'm wondering if we even need to check whether the routing is valid or not -- do we need to add a continue here, or does it "just work" without it as well?

I think it should "just work", given it was possible for `SyncTokenAndMaybeUpdateCache` to invalidate the routing before your change as well and we weren't checking whether the routing was valid or not here.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2724 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NruwMVG2Nzo_agrg7Cl:-NruwMVG2Nzo_agrg7Cm:bh006ag) ([raw file](https://github.com/cockroachdb/cockroach/blob/903e2fc5e8286ecff775bd7407bd399889ca97ce/pkg/kv/kvclient/kvcoord/dist_sender.go#L2724)):*
> ```Go
> 					// may reject the request with a NotLeaseHolderError if it does not
> 					// have a sufficient closed timestamp. In response, we should
> 					// immediately redirect to the leaseholder, without a backoff
> ```

This commentary is still worth keeping around.
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1460 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqGkkK8Ya9TN6M33zu:-NrusMPZCKHJn5qe8C-m:b-8b1kp6) ([raw file](https://github.com/cockroachdb/cockroach/blob/505dcae125dd4b24758f87968b405b90dd4d7b75/pkg/kv/kvclient/rangecache/range_cache.go#L1460)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

I think I can remove the nil rangeDesc check, but I can't remove the Lease check below.
</blockquote></details>

Ack. 
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1487 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqHgaD5-zosqmAHJKF:-NruwVjiDeiBZ1zDHU7E:bcvzki4) ([raw file](https://github.com/cockroachdb/cockroach/blob/505dcae125dd4b24758f87968b405b90dd4d7b75/pkg/kv/kvclient/rangecache/range_cache.go#L1487)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

The idea was to allow calling this method when we don't have a new Lease to provide. I agree that it makes the conditions for this a little more subtle and I don't love it, but the alternative was to have seperate methods for `SyncLease` and `SyncRangeDescriptor` Maybe we should talk briefly on this.
</blockquote></details>

See my comment above about the practical cases where we're calling this method with no lease provided. If this is going to change with the partial partitioning work, then I think the change you have here makes sense -- it's not great that we're expanding the possible inputs here, but we have no choice. 

I think this is a better approach than having separate methods here. Mostly because the descriptor and lease go hand in hand, and we have special case handling for when the lease and descriptor aren't compatible. If we were to tease these apart, we'd have to push some ordering on to the caller (e.g. first update the descriptor, then the lease) which isn't ideal. 


<!-- Sent from Reviewable.io -->

andrewbaptist reviewed (commented):
The tests in the partial partition work do somewhat cover this case, but more specifically the goal was to handle the NLHE more robustly in all cases. Previously we inefficiently handled some cases, but it was OK because we made some progress. I can try and see if there is a way to add a test for this, but otherwise let me know if its OK to just add a TODO and come back to this.

Lets chat on Monday.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-:-NrvsC_Z4WqOUhk9JDR_:b-bbt8yk)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2670 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqJXdZ99CPKQDINyfD:-NrvsYsX-qFKfNllmjmB:bti282s) ([raw file](https://github.com/cockroachdb/cockroach/blob/903e2fc5e8286ecff775bd7407bd399889ca97ce/pkg/kv/kvclient/kvcoord/dist_sender.go#L2670)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

> A request comes in for Range 1 on a non-leaseholder who doesn't know the new leaseholder.

Getting a nil lease here is extremely rare (I'm looking at the callers of `NewNotLeaseHolderError`). So there aren't many practical cases where we'd have both a nil lease and a stale range descriptor. Do we think the theoretical benefit here is worth the added complexity?
</blockquote></details>

The goal was to more simply pass the full results of the NLHE from the server back into the client. I was hoping this was a simplification rather than added complexity. The NLHE had grown in complexity over time and the code no longer reflected what it was doing. 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2676 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqJyx1BuI1-0DErr25:-NrvswEN0dqan8e529IQ:b-i8tii1) ([raw file](https://github.com/cockroachdb/cockroach/blob/903e2fc5e8286ecff775bd7407bd399889ca97ce/pkg/kv/kvclient/kvcoord/dist_sender.go#L2676)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Sorry for not being clear, this was contingent on my previous comment of keeping the old structure -- the if condition here isn't necessary (on master) because it's already inside the same if condition above.
</blockquote></details>

On master it was not handling this case well. It would throw away the NLHE even though it had some updated information. I wanted to reduce special cases by simply passing the "whole" NLHE error into SyncToken rather than try and determine externally if it was approporiate.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2683 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqK8zl6eBtgD9IIMaS:-NrvtT5p49tJlXJ8bZB6:bohli2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/22af9b4fbf7095f22da55da414285e6c8a05a645/pkg/kv/kvclient/kvcoord/dist_sender.go#L2683)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

We're now assuming that the only error here means the routing is invalid, which might not hold true forever. One pattern I've seen used before is to return a special type of error (say `routingInvalidError`), check for that, and only continue in that case. 

However, for this, it might just be easier to check if the routing is valid or not, instead of introducing a new error type. Alternatively, I'm wondering if we even need to check whether the routing is valid or not -- do we need to add a continue here, or does it "just work" without it as well?

I think it should "just work", given it was possible for `SyncTokenAndMaybeUpdateCache` to invalidate the routing before your change as well and we weren't checking whether the routing was valid or not here.
</blockquote></details>

You're right that we didn't need to check if the routing is valid or not. To simplify this. I just check if there is an error and log it rather than doing anything with it. We can add different types if we need to handle this differently in the future. 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2724 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NruwMVG2Nzo_agrg7Cl:-Nrvwuc18C-k4YLSv9XC:b-t5v5b4) ([raw file](https://github.com/cockroachdb/cockroach/blob/903e2fc5e8286ecff775bd7407bd399889ca97ce/pkg/kv/kvclient/kvcoord/dist_sender.go#L2724)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

This commentary is still worth keeping around.
</blockquote></details>

That was accidental. I moved this back. 
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1487 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqHgaD5-zosqmAHJKF:-Nrvx0Oi7AR5fEDgY19l:bsk2rro) ([raw file](https://github.com/cockroachdb/cockroach/blob/505dcae125dd4b24758f87968b405b90dd4d7b75/pkg/kv/kvclient/rangecache/range_cache.go#L1487)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

See my comment above about the practical cases where we're calling this method with no lease provided. If this is going to change with the partial partitioning work, then I think the change you have here makes sense -- it's not great that we're expanding the possible inputs here, but we have no choice. 

I think this is a better approach than having separate methods here. Mostly because the descriptor and lease go hand in hand, and we have special case handling for when the lease and descriptor aren't compatible. If we were to tease these apart, we'd have to push some ordering on to the caller (e.g. first update the descriptor, then the lease) which isn't ideal.
</blockquote></details>

I understand your hesitancy but in reality there is a real valid case where we learned new range information but don't have corresponding lease information even without the partial partition work. I agree we can't use separate methods because we would want to change both the lease and the replica under the same lock. From a "style" perspective, this method would probably look better if it took a `roachpb.RangeInfo` rather than separate objects, and then it wouldn't seem quite so blurry what it is doing. I added a TODO to consider changing it to doing that. In practice, we should really change the NLHE to have a RangeInfo as well as it would be a little bit better. 

In partial partitions the number of possible combinations of errors is increased, so it is simpler to just handle all the possible inputs well rather than thinking about how common the edge cases are or not. This is a real use that we have and incorrectly handle today because of the existing contract/signature. 


<!-- Sent from Reviewable.io -->

andrewbaptist commented:
Hold off reviewing - I need to look at `TestReplicaCircuitBreaker_Partial_Retry` as one of the checks succeeds with this change where it would previously fail. Specifically 
```
Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_replica_circuit_breaker_test.go:796
github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_replica_circuit_breaker_test.go:850
        Error:      	An error is expected but got nil.
        Test:       	TestReplicaCircuitBreaker_Partial_Retry
```

andrewbaptist commented:
@arulajmani I updated `TestNoBackoffOnNotLeaseHolderErrorWithoutLease` to verify that if there is a request without a lease, but with a newer range descriptor, that we now see that update. This test change also validates that we can call the NLHE with a nil lease without problems (it was always calling it that way, but now verifies that it updates the cache as expected). Previously it would simply ignore the NLHE.

arulajmani reviewed (commented):
> Reviewed 1 of 5 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119747)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @andrewbaptist)

___
*[`-- commits` line 33 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NstAV2F0wGL7hor7mUK):*
See my comment below, but this isn't fairly unusual.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2670 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqJXdZ99CPKQDINyfD:-NsV-EDhBFr_5zdiiprl:bh7k714) ([raw file](https://github.com/cockroachdb/cockroach/blob/903e2fc5e8286ecff775bd7407bd399889ca97ce/pkg/kv/kvclient/kvcoord/dist_sender.go#L2670)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

The goal was to more simply pass the full results of the NLHE from the server back into the client. I was hoping this was a simplification rather than added complexity. The NLHE had grown in complexity over time and the code no longer reflected what it was doing.
</blockquote></details>

By added complexity I meant the increased testing area and the set of inputs the function considers valid. This in contrast to simply saying that if a NLHE doesn't contain leaseholder information (which should be rare) we should just move on to the next replica. 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2712 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-Nst9wR-9Vlg_oSMufmJ:-Nst9wR-9Vlg_oSMufmK:bnoxjee) ([raw file](https://github.com/cockroachdb/cockroach/blob/114617af7ee17c7f6b74c60b7177a60fe54e6581/pkg/kv/kvclient/kvcoord/dist_sender.go#L2712)):*
> ```Go
> 				// If the range bounds changed our routing may be invalid.
> 				if err := routing.SyncTokenAndMaybeUpdateCache(ctx, errLease, errDesc); err != nil {
> 					log.VErrEventf(ctx, 1, "problem syncing our token: %s", err)
> ```

I'm not sure about this logging, and this commit in general. Getting information in an NLHE that invalidates the token is legit -- why are we treating it as an error? Separately, this logging sounds more alarming than it should be. 

I'd be in favour of dropping this commit entirely.
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1487 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqHgaD5-zosqmAHJKF:-NsV0Ko543bDWKZDL8mG:b-lze00w) ([raw file](https://github.com/cockroachdb/cockroach/blob/505dcae125dd4b24758f87968b405b90dd4d7b75/pkg/kv/kvclient/rangecache/range_cache.go#L1487)):*
> In practice, we should really change the NLHE to have a RangeInfo as well as it would be a little bit better.

IMO this isn't worth the backwards compatibility complexity cost we'd have to pay. 

> In partial partitions the number of possible combinations of errors is increased, so it is simpler to just handle all the possible inputs well rather than thinking about how common the edge cases are or not.

Sounds good to me. 
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1484 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-Nst6uoWBVHnuHdf3UeQ:-Nst6uoWBVHnuHdf3UeR:b-imlzzb) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ebf018908c89cbe4045df6fccc76251b434b9f1/pkg/kv/kvclient/rangecache/range_cache.go#L1484)):*
> ```Go
> 	// will be replaced by the supplied lease if the associated range descriptor
> 	// is non-stale.
> 	// TODO(baptist): A speculative lease shouldn't always replace a
> ```

Is this true? We return speculative leases when we decide not to acquire a lease because we're not the raft leader and the raft leader is alive. 

I may be missing something, and if I am, this deserves more than a TODO -- can you flesh out the details here and open a GitHub issue instead?
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1495 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-Nst8S1pEsfJDpfoPznO:-Nst8S1pEsfJDpfoPznP:bepuneh) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ebf018908c89cbe4045df6fccc76251b434b9f1/pkg/kv/kvclient/rangecache/range_cache.go#L1495)):*
> ```Go
> 	if !l.Empty() &&
> 		(neitherSpeculative && l.Sequence > e.lease.Sequence ||
> 			eitherSpeculative && rangeDesc.Generation >= e.desc.Generation ||
> ```

Nice readability change! 


<!-- Sent from Reviewable.io -->

andrewbaptist reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119747)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @arulajmani)

___
*[`-- commits` line 33 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NstAV2F0wGL7hor7mUK):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

See my comment below, but this isn't fairly unusual.
</blockquote></details>

I ended up dropping this commit.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2670 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqJXdZ99CPKQDINyfD:-Nstx2ES4MAvYzFWLnFD:b-djd8sc) ([raw file](https://github.com/cockroachdb/cockroach/blob/903e2fc5e8286ecff775bd7407bd399889ca97ce/pkg/kv/kvclient/kvcoord/dist_sender.go#L2670)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

By added complexity I meant the increased testing area and the set of inputs the function considers valid. This in contrast to simply saying that if a NLHE doesn't contain leaseholder information (which should be rare) we should just move on to the next replica.
</blockquote></details>

My concern is still how we handle cases where we get partial information from the server. I'm not sure there is a specific case right now where this matters, but I wanted to plug any potential holes with us being stuck looping or failing due to incomplete handling of this request. 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2712 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-Nst9wR-9Vlg_oSMufmJ:-Nsu3M5Q8wqC2prjEKJo:beekw35) ([raw file](https://github.com/cockroachdb/cockroach/blob/114617af7ee17c7f6b74c60b7177a60fe54e6581/pkg/kv/kvclient/kvcoord/dist_sender.go#L2712)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I'm not sure about this logging, and this commit in general. Getting information in an NLHE that invalidates the token is legit -- why are we treating it as an error? Separately, this logging sounds more alarming than it should be. 

I'd be in favour of dropping this commit entirely.
</blockquote></details>

I dropped this commit.
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1487 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqHgaD5-zosqmAHJKF:-Nsty-q9167bq2dNQR_P:bh5b5rm) ([raw file](https://github.com/cockroachdb/cockroach/blob/505dcae125dd4b24758f87968b405b90dd4d7b75/pkg/kv/kvclient/rangecache/range_cache.go#L1487)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

> In practice, we should really change the NLHE to have a RangeInfo as well as it would be a little bit better.

IMO this isn't worth the backwards compatibility complexity cost we'd have to pay. 

> In partial partitions the number of possible combinations of errors is increased, so it is simpler to just handle all the possible inputs well rather than thinking about how common the edge cases are or not.

Sounds good to me.
</blockquote></details>

Agreed - I'm not planning to change this handling now since it would require careful upgrade handling. Logically I'm trying to get to a point where the RangeCache deals with `RangeInfo` only (for both values and syncing). There is some complexity around this related to speculative leases and descriptors, which is a big reason not to make this change. I remove any comments related to making this change.
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1484 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-Nst6uoWBVHnuHdf3UeQ:-Nstz2x-C-Li4BCn7R67:blmo7kg) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ebf018908c89cbe4045df6fccc76251b434b9f1/pkg/kv/kvclient/rangecache/range_cache.go#L1484)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Is this true? We return speculative leases when we decide not to acquire a lease because we're not the raft leader and the raft leader is alive. 

I may be missing something, and if I am, this deserves more than a TODO -- can you flesh out the details here and open a GitHub issue instead?
</blockquote></details>

You are correct about the only case where we return a speculative lease today which is in replica_proposal_buf in `maybeRejectUnsafeProposalLocked`. My TODO has to do with the case where the contract of the NLHE is unclear if we change any places on the server where we return a NLHE with a speculative lease. Specifically the "contract" of the range cache is that it always moves forward, but the current handling violates this in the case of speculative leases. It would be much better if this was a "check and set" type operation where we could only propose a speculative lease by sending the lease sequence of the lease that our speculation was based on. 

The concerning scenario I'm thinking of is where a node knows the raft leader, but is far behind and doesn't know who the current leaseholder is. In general knowing the raft leader doesn't mean you have applied the entire raft log and there is no time bound no how long this could take to apply. If you are behind and keep returning speculative lease entries based on stale information, then you can cause problems in the case of a prolonged unresolved leader/leaseholder split. So the TODO is more about the fact that the range cache shouldn't make assumptions about which cases are valid to return speculative leases.

Also until we figure out how to handle #120150 which is related to the case of sending a speculative lease error back when we might not know the leader. 

In some ways I think we might just want to kill the concept of speculative leases as they add a fair bit of complexity for a very small gain. We would figure out the same information by trying other replicas and it would likely at best help remove one retry in a very edge case.
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1495 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-Nst8S1pEsfJDpfoPznO:-NstyUDP3aPgnNzgauA1:b-zcjvte) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ebf018908c89cbe4045df6fccc76251b434b9f1/pkg/kv/kvclient/rangecache/range_cache.go#L1495)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Nice readability change!
</blockquote></details>

Thanks


<!-- Sent from Reviewable.io -->

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 4 of 4 files at r16, 2 of 2 files at r17, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-:-NsyFNV0BHQ7fG9UrhMi:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @abarganier and @andrewbaptist)

___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1487 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-NrqHgaD5-zosqmAHJKF:-NsyDWWJ1TS9kif8gm6R:b-wtet1v) ([raw file](https://github.com/cockroachdb/cockroach/blob/505dcae125dd4b24758f87968b405b90dd4d7b75/pkg/kv/kvclient/rangecache/range_cache.go#L1487)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Agreed - I'm not planning to change this handling now since it would require careful upgrade handling. Logically I'm trying to get to a point where the RangeCache deals with `RangeInfo` only (for both values and syncing). There is some complexity around this related to speculative leases and descriptors, which is a big reason not to make this change. I remove any comments related to making this change.
</blockquote></details>

Ack, thanks for removing the commentary related to this change.
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1484 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/119747#-Nst6uoWBVHnuHdf3UeQ:-NsyDeqy5G3r_bOt6UkU:b-1ouhxv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ebf018908c89cbe4045df6fccc76251b434b9f1/pkg/kv/kvclient/rangecache/range_cache.go#L1484)):*
> The concerning scenario I'm thinking of is where a node knows the raft leader, but is far behind and doesn't know who the current leaseholder is. In general knowing the raft leader doesn't mean you have applied the entire raft log and there is no time bound no how long this could take to apply. If you are behind and keep returning speculative lease entries based on stale information, then you can cause problems in the case of a prolonged unresolved leader/leaseholder split.

This all makes sense, but doesn't it point to speculative leases being independent of lease information in the replica's raft log? It's speculative, in that in the common case the replica returning the lease expects the raft leader to be the leaseholder as well. 

I also don't see how we can be in the "keep returning speculative leases" scenario -- we'll move on to the raft leader, which will either be the leaseholder or point to a different replica if there's a leader/leaseholder split. Either way, the replica which returned the speculative lease can't be the leaseholder because we take care not to transfer leases to replicas that are behind.

I think getting rid of the TODO here makes sense. I may still be missing something, so let's open a Github issue to capture all our thoughts if we want to make some changes here. Ditto for if we want to remove speculative leases entirely (IMO that deserves a github issue too before we make the change).


<!-- Sent from Reviewable.io -->

andrewbaptist commented:
TFTR!

bors r=arulajmani

Opening an issue related to this now.
